ext/storage_sources/s3_store/s3_connection.cpp

```diff
@@ -11,51 +11,78 @@
 
 /*
  * ListBuckets --
- *     Builds a list of buckets from AWS account into a vector. Returns true if success, otherwise
- *     false.
+ *     Builds a list of buckets from AWS account into a vector. Returns 0 if success, otherwise
+ *     1.
  */
-bool
+int
 S3Connection::ListBuckets(std::vector<std::string> &buckets) const
 {
     auto outcome = m_S3CrtClient.ListBuckets();
     if (outcome.IsSuccess()) {
         for (const auto &bucket : outcome.GetResult().GetBuckets())
             buckets.push_back(bucket.GetName());
-        return true;
+        return (0);
     } else {
         std::cerr << "Error in ListBuckets: " << outcome.GetError().GetMessage() << std::endl
                   << std::endl;
-        return false;
+        return (1);
     }
 }
 
 /*
  * ListObjects --
- *     Builds a list of object names from a S3 bucket into a vector. Returns true if success,
- *     otherwise false.
+ *     Builds a list of object names from an S3 bucket into a vector. Returns 0 if success,
+ * otherwise 1.
  */
-bool
-S3Connection::ListObjects(const std::string &bucketName, std::vector<std::string> &objects) const
+int
+S3Connection::ListObjects(const std::string &bucketName, const std::string &prefix,
+  std::vector<std::string> &objects, uint32_t &countp, uint32_t nPerIter, uint32_t maxObjects) const
 {
-    Aws::S3Crt::Model::ListObjectsRequest request;
-    request.WithBucket(bucketName);
-    Aws::S3Crt::Model::ListObjectsOutcome outcomes = m_S3CrtClient.ListObjects(request);
+    objects.clear();
+    Aws::S3Crt::Model::ListObjectsV2Request request;
+
+    request.SetBucket(bucketName);
+    request.SetPrefix(prefix);
+    request.SetMaxKeys(nPerIter);
+
+    countp = 0;
+    if (maxObjects != 0 && (maxObjects - countp) < nPerIter)
+        request.SetMaxKeys(maxObjects - countp);
+
+    Aws::S3Crt::Model::ListObjectsV2Outcome outcomes = m_S3CrtClient.ListObjectsV2(request);
 
     if (outcomes.IsSuccess()) {
-        for (const auto &object : outcomes.GetResult().GetContents())
+        auto result = outcomes.GetResult();
+        std::string continuationToken = result.GetNextContinuationToken();
```

Comment:
You can create the `continuationToken` right before it is used, just before the next `while` loop.